File: CPP/DFS/797. All Paths From Source to Target.cpp
class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        int target = graph.size()-1;
        vector<vector<int>> res;
        
        dfs(graph, 0, {}, target, res);
        return res;
    }

    void dfs(vector<vector<int>>& graph, int pos, vector<int> path, int& target, vector<vector<int>>& res){
        path.push_back(pos);
        if(pos == target){
            res.push_back(path);
        }
        else{
            for (int next_pos: graph[pos]){
                dfs(graph, next_pos, path, target, res);
            }
        }
        return;
    }
};

File: Python/DFS/797. All Paths From Source to Target.py
from typing import List

class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        target = len(graph)-1

        res = []
        def dfs(pos, curPath):
            if pos == target:
                res.append(curPath.copy())
                return
            
            for n in graph[pos]:
                dfs(n, curPath+[n])
        
        dfs(0,[0])

        return res



if __name__ == '__main__':
    graph = [[4,3,1],[3,2,4],[3],[4],[]]
    sol = Solution()
    print(sol.allPathsSourceTarget(graph))